smolyay/samples.py

```diff
@@ -0,0 +1,207 @@
+        num_points : int
+            The number of points to generate
+
+        Returns
+        -------
+        self
+            The UnidimensionalPointSet
+        """
+        if num_points == 0:
+            self._points = []
+        degree = num_points - 1
+        if degree > 0:
+            self._points = list(
+                -numpy.cos(numpy.pi * numpy.linspace(0, degree, degree + 1) / degree)
+            )
+        else:
+            self._points = [0]
+        return self
+
+
+class NestedClenshawCurtisPointSet(ClenshawCurtisPointSet):
+    r"""Generate nested Clenshaw Curtis points in accordance with a growth rule
+
+    The :attr:`points` for this interpolation scheme are the extrema of the
+    Chebyshev polynomials of the first kind on the domain :math:`[-1, 1]`:
+
+    .. math::
+
+        x_i^* = -\cos(\pi i/n), i = 0,...,n
+
+    For the special case :math:`n = 0`, there is only one point :math:`x_0^* = 0`.
+
+    The nested Clenshaw Curtis points come from the nested extrema of the
+    Chebyshev polynomials of the first kind :math:0, 2, 2^{k} where k is an
+    integer.
+
+    These extrema are symmetric about 0. When generating levels of points, each
+    level with contain a certain number of symmetric pairs except the first which
+    also includes 0. The number of points assigned to each level can vary.
+    The default, exponential growth adds points such that the new extrema
+    calculated from a Chebyshev polynomial 2^{k} is given its own level k-1.
+    Alternative growth rules may delay the addition of new extrema and have
+    intermediary empty levels, but the general order that points are added to
+    levels remains the same.
+
+    This class will generate the nested extrema in the order that corresponds
+    to the exponential growth. For increasing k, the unique extrema of Chebyshev
+    polynomial with degree 2^{k - 1} will be added followed by the extrema of the
+    polynomial with degree 2^{k} , then 2^{k + 1} , then 2^{k + 2} and so on.
+    """
+
+    def generate_points(self, num_points):
+        r"""Generating the points
+
+        Parameters
+        ----------
+        num_points : int
+            The number of points to generate
+
+        Returns
+        -------
+        self
+            The UnidimensionalPointSet
+        """
+        points = [0]
+        degree = 0
+        counter = 0
+        while len(points) < num_points:
+            counter = counter + 1
+            degree = 2**counter
+            if counter == 1:
+                indexes = numpy.linspace(0, degree, 2, dtype=int)
+            else:
+                indexes = numpy.linspace(1, degree - 1, degree - 1, dtype=int)
+                indexes = indexes[~(numpy.gcd(indexes, degree) > 1)]
+            new_points = list(-numpy.cos(numpy.pi * indexes / degree))
+            points.extend(new_points)
+        self._points = points[:num_points]
+        return self
+
+
+class TrigonometricPointSet(UnidimensionalPointSet):
```

Comment:
Similar changes are needed for these classes as for the Clenshaw-Curtis ones. There also should be a nested and unnested version of the trig points.

Reply:
With the addition of the `degree` parameter for nonnested point sets and `max_level` for the nested point sets, the Clenshaw-Curtis and Trigonometric sets no longer inherit anything from each other.

Perhaps there should be an abstract class ClenshawCurtis that sets the domain property that both the Clenshaw-Curtis sets can inherit from, and the same thing with the Trigonometric sets.

There could also be a NestedPointSet that is an abstract class for any point set that uses the `max_level` parameter, and a NonNestedPointSet for any point set that uses the `degree` parameter. It's possible for Python classes to have multiple parents, so it would be nice to have the point set objects to clearly demonstrate how they are related via their inheritance structure.

Reply:
`degree` and `max_level` are parameters of the constructors for these concrete classes, but I don't think they should be required as part of the abstract base class interfaces. I probably wouldn't add another base class just to avoid defining the domain property twice.

If you want to use more inheritance, the nested classes should still be able to use the non-nested constructor to figure out all the points by computing the degree from the max level, right? However, then there is still the issue of assigning points to levels, and I would assume that would best be done at the same time as the points are made, so that probably limits the usefulness of this.